### PR TITLE
CIRC-1051 Receiving overdue notices after item claimed returned

### PR DESCRIPTION
**Pull request by Illia-Dolzhenko**
## Purpose
- Stop sending scheduled notices after item claimed returned
## Approach
- Check if item is claimed returned before sending notice

Resolves: [CIRC-1051](https://issues.folio.org/browse/CIRC-1051)
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.
- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: src/test/java/api/loans/ClaimItemReturnedAPITests.java [hunk 1]
```diff
@@ -41,8 +49,34 @@
...
+      .withAfterTiming(Period.minutes(5))
+      .recurring(Period.minutes(5))
+      .sendInRealTime(true)
+      .create();
+
+    NoticePolicyBuilder noticePolicyBuilder = new NoticePolicyBuilder()
+      .withName("loan policy")
+      .withLoanNotices(Collections.singletonList(loanNotice));
+
+    use(noticePolicyBuilder);
```
Comment (roman-barannyk) [inline src/test/java/api/loans/ClaimItemReturnedAPITests.java:78]: Does it make sense keep it in @Before method if you use it only for one test?

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 2]
```diff
@@ -228,7 +229,7 @@ public static LoanScheduledNoticeHandler using(Clients clients, DateTime systemT
     Loan loan = relatedRecords.getLoan();
     ScheduledNoticeConfig noticeConfig = notice.getConfiguration();
 
-    if (loan.isClosed() || !noticeConfig.isRecurring()) {
+    if (loan.isClosed() || !noticeConfig.isRecurring() || noticeIsNotRelevant(notice, loan)) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:232]: Remove first check: `loan.isClosed()`

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 3]
```diff
@@ -250,7 +251,31 @@ public static LoanScheduledNoticeHandler using(Clients clients, DateTime systemT
   }
 
   public boolean noticeIsNotRelevant(ScheduledNotice notice, Loan loan) {
-    return loan.isClosed() || beforeDueDateNoticeIsNotRelevant(notice, loan);
+    if (loan.isClosed()) {
+      return true;
+    }
+    TriggeringEvent triggeringEvent = notice.getTriggeringEvent();
+    if (triggeringEvent == DUE_DATE) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:258]: Try switch

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 4]
```diff
@@ -250,7 +251,31 @@ public static LoanScheduledNoticeHandler using(Clients clients, DateTime systemT
...
+      return dueDateNoticeIsNotRelevant(notice, loan);
+    } else if (triggeringEvent == AGED_TO_LOST) {
+      return agedToLostNoticeIsNotRelevant(loan);
+    }
+    log.error("Unexpected triggering event {}", triggeringEvent.getRepresentation());
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:263]: Extract error message into a variable

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 5]
```diff
@@ -250,7 +251,31 @@ public static LoanScheduledNoticeHandler using(Clients clients, DateTime systemT
...
+    throw new UnsupportedOperationException("Unexpected triggering event: " +
+      triggeringEvent.getRepresentation());
+  }
+
+  private boolean dueDateNoticeIsNotRelevant(ScheduledNotice notice, Loan loan) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:268]: Please concider logging the reason for irrelevance

File: src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java [hunk 6]
```diff
@@ -192,6 +198,57 @@ public void shouldStopSendingAgedToLostNoticesOnceLostItemFeeWasCharged() {
...
+      TIMING_PERIOD.timePeriod());
+    scheduledNoticeProcessingClient.runLoanNoticesProcessing(
+      firstRunTime.plus(RECURRENCE_PERIOD.timePeriod()).plusMinutes(1));
+
+    Awaitility.await()
+      .atMost(1, TimeUnit.SECONDS)
+      .until(scheduledNoticesClient::getAll, empty());
+  }
+
+  private AgeToLostResult createAgedToLostNotice() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java:233]:
```suggestion
  private AgeToLostResult createRecurringAgedToLostNotice() {
```

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java [hunk 7]
```diff
@@ -22,19 +26,25 @@
...
 import api.support.http.IndividualResource;
+import api.support.http.ItemResource;
 import io.vertx.core.json.JsonObject;
 
 public class DueDateScheduledNoticesTests extends APITests {
   private static final String BEFORE_TIMING = "Before";
   private static final String UPON_AT_TIMING = "Upon At";
   private static final String AFTER_TIMING = "After";
+  private static final Period TIMING_AFTER = Period.minutes(5);
+  private static final Period TIMING_RECURRING = Period.minutes(10);
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesTests.java:47]:
```suggestion
  private static final Period TIMING_PERIOD_AFTER = Period.minutes(5);
  private static final Period TIMING_PERIOD_RECURRING = Period.minutes(10);
```

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java [hunk 8]
```diff
@@ -621,4 +631,138 @@ public void scheduledOverdueNoticesShouldBeDeletedAfterOverdueFineIsCharged() {
...
+      .claimItemReturned(new ClaimItemReturnedRequestBuilder()
+        .forLoan(loan.getId().toString())
+        .withItemClaimedReturnedDate(now()));
+
+    var dueDate = getDateTimeProperty(loan.getJson(), "dueDate");
+    scheduledNoticeProcessingClient.runLoanNoticesProcessing(dueDate.plusDays(1));
+
+    Awaitility.await()
+      .atMost(1, TimeUnit.SECONDS)
+      .until(scheduledNoticesClient::getAll, empty());
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesTests.java:667]: Check that patron notice was not sent. See if it's possible to implement these tests synchronously.

File: src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java [hunk 9]
```diff
@@ -192,6 +198,57 @@ public void shouldStopSendingAgedToLostNoticesOnceLostItemFeeWasCharged() {
...
+  public void shouldStopSendingAgedToLostNoticesOnceItemIsDeclaredLost() {
+    AgeToLostResult agedToLostLoan = createAgedToLostNotice();
+
+    declareLostFixtures.declareItemLost(agedToLostLoan.getLoan().getJson());
+    final DateTime firstRunTime = getAgedToLostDate(agedToLostLoan).plus(
+      TIMING_PERIOD.timePeriod());
+    scheduledNoticeProcessingClient.runLoanNoticesProcessing(
+      firstRunTime.plus(RECURRENCE_PERIOD.timePeriod()).plusMinutes(1));
+
+    Awaitility.await()
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java:211]: Check that notice was not sent

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java [hunk 10]
```diff
@@ -621,4 +631,138 @@ public void scheduledOverdueNoticesShouldBeDeletedAfterOverdueFineIsCharged() {
...
+
+  @Test
+  public void scheduledOverdueNoticesShouldBeDeletedAfterDeclaredLost() {
+    NoticePolicyBuilder noticePolicy = createSendNoticePolicy();
+    use(noticePolicy);
+    IndividualResource loan = createScheduledNotice(now());
+
+    declareLostFixtures.declareItemLost(loan.getJson());
+    var dueDate = getDateTimeProperty(loan.getJson(), "dueDate");
+    scheduledNoticeProcessingClient.runLoanNoticesProcessing(dueDate.plusDays(1));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesTests.java:678]: Try setting smaller delay, like 1 second

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 11]
```diff
@@ -250,7 +251,31 @@ public static LoanScheduledNoticeHandler using(Clients clients, DateTime systemT
...
+    return beforeDueDateNoticeIsNotRelevant(notice, loan)
+      || loan.getItem().isDeclaredLost()
+      || loan.getItem().getStatus() == ItemStatus.AGED_TO_LOST
+      || loan.isRenewed()
+      || loan.getItem().isClaimedReturned()
+      || (loan.hasDueDateChanged() && loan.getDueDate().isAfter(systemTime));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:274]: Consider removing this condition, it's already covered in `ChangeDueDateResource`

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java [hunk 12]
```diff
@@ -621,4 +631,138 @@ public void scheduledOverdueNoticesShouldBeDeletedAfterOverdueFineIsCharged() {
...
+      .withLoanNotices(Collections.singletonList(loanNotice));
+    use(noticePolicy);
+    return noticePolicy;
+  }
+
+  private IndividualResource createScheduledNotice(DateTime loanDate) {
+    return createScheduledNotice(itemsFixture.basedUponNod(), usersFixture.steve(), loanDate);
+  }
+
+  private IndividualResource createScheduledNotice(ItemResource item, IndividualResource user,
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesTests.java:753]: Think of a better name. This method itself does not create notices.

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java]: done

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 3 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java]: updated

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 4 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java]: done

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 5 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:271]: added

File: src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java (hunk 6 above)
Comment (roman-barannyk) [reply src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java]: renamed

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java (hunk 8 above)
Comment (roman-barannyk) [reply src/test/java/api/loans/DueDateScheduledNoticesTests.java]: updated

File: src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java (hunk 9 above)
Comment (roman-barannyk) [reply src/test/java/api/loans/AgedToLostScheduledNoticesProcessingTests.java:211]: added

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java (hunk 10 above)
Comment (roman-barannyk) [reply src/test/java/api/loans/DueDateScheduledNoticesTests.java]: 1 hour is possible as min

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java (hunk 12 above)
Comment (roman-barannyk) [reply src/test/java/api/loans/DueDateScheduledNoticesTests.java]: renamed

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 13]
```diff
@@ -250,7 +251,73 @@ public static LoanScheduledNoticeHandler using(Clients clients, DateTime systemT
...
+    switch (triggeringEvent) {
+      case DUE_DATE:
+        return dueDateNoticeIsNotRelevant(notice, loan);
+      case AGED_TO_LOST:
+        return agedToLostNoticeIsNotRelevant(loan);
+      default:
+        var errorMessage = String.format("Unexpected triggering event %s",
+          triggeringEvent.getRepresentation());
+        log.error(errorMessage);
+        throw new UnsupportedOperationException(errorMessage);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:264]: Maybe `IllegalArgumentException` is more appropriate here?

File: src/test/java/api/loans/DueDateScheduledNoticesTests.java (hunk 10 above)
Comment (OleksandrVidinieiev) [reply src/test/java/api/loans/DueDateScheduledNoticesTests.java]: see method `DateTime#plusSeconds`

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 5 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:268]: I think this method can shortened to something like this:
```
  private boolean dueDateNoticeIsNotRelevant(ScheduledNotice notice, Loan loan) {
    String irrelevanceReason = null;
    
    if (beforeDueDateNoticeIsNotRelevant(notice, loan)) {
      irrelevanceReason = "foo";
    } else if (loan.isDeclaredLost()) {
      irrelevanceReason = "bar";
    } else if (loan.isRenewed()) {
      ...
    } else ...

    if (irrelevanceReason != null) {
      log.warn("Scheduled patron notice {} is irrelevant: {}", notice.getId(), irrelevanceReason);
      return true;
    }
    
    return false;
  }
```

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java [hunk 14]
```diff
@@ -266,55 +268,58 @@ public boolean noticeIsNotRelevant(ScheduledNotice notice, Loan loan) {
   }
 
   private boolean dueDateNoticeIsNotRelevant(ScheduledNotice notice, Loan loan) {
+    List<String> logMessages = new ArrayList<>();
     if (beforeDueDateNoticeIsNotRelevant(notice, loan)) {
-      log.warn("The notice {} is irrelevant. The due date is before", notice.getId());
-      return true;
+      logMessages.add(String.format("The notice %s is irrelevant. The due date " +
+          "of loan %s is before", notice.getId(), loan.getId()));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:274]: Before what?

**Review by OleksandrVidinieiev — APPROVED** [review]

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 14 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java]: before now, updated

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java (hunk 5 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeHandler.java:270]: Changed this implementation with list of messages

**Pull request merged**